### DISCUSSION TITLE
Keep V code in generated files

### DISCUSSION THREAD

**Discussion by duarteroso** [vlang/v#11722] [Ideas]
### Description

Something I wanted to bring here since I started using the generated C code to debug stuff.

As you (might) know, Unity converts C# to C++ using their tool called `IL2CPP`.
The cool thing `IL2CPP` does to help debugging post-generation is to keep the original C# as comments in the generated code.

### Example

Bellow you can see an example of a C# script and it's `IL2CPP` translation. There is a lot of garbage in the middle but the original code can still be searched for. Very helpful when debugging.

```csharp
public class MyScript : MonoBehaviour
{
    // Start is called before the first frame update
    void Start()
    {
        Debug.Log("This a log");
        int i = 0;
        if(i == 0)
        {
            string s = string.Empty;
            s = "some" + "thing";
        }
    }
}
```

```cpp
// System.Void MyScript::Start()
IL2CPP_EXTERN_C IL2CPP_METHOD_ATTR void MyScript_Start_m5193661BFC5141BBCC8C4C551D045E7309C0BA1B (MyScript_t40D8B6B82135E49C862E290A0A7DFA245E333969 * __this, const RuntimeMethod* method)
{
	static bool s_Il2CppMethodInitialized;
	if (!s_Il2CppMethodInitialized)
	{
		il2cpp_codegen_initialize_method (MyScript_Start_m5193661BFC5141BBCC8C4C551D045E7309C0BA1B_MetadataUsageId);
		s_Il2CppMethodInitialized = true;
	}
	{
		// Debug.Log("This a log");
		IL2CPP_RUNTIME_CLASS_INIT(Debug_tEB68BCBEB8EFD60F8043C67146DC05E7F50F374B_il2cpp_TypeInfo_var);
		Debug_Log_mC26E5AD0D8D156C7FFD173AA15827F69225E9DB8(_stringLiteral5C95FD8CBC53CCADA3D83F40A5C95812F7985FD5, /*hidden argument*/NULL);
		// int i = 0;
		// if(i == 0)
		if (0)
		{
			goto IL_001d;
		}
	}
	{
		// string s = string.Empty;
		String_t* L_0 = ((String_t_StaticFields*)il2cpp_codegen_static_fields_for(String_t_il2cpp_TypeInfo_var))->get_Empty_5();
		// s = "some" + "thing";
		// Debug.Log(s);
		IL2CPP_RUNTIME_CLASS_INIT(Debug_tEB68BCBEB8EFD60F8043C67146DC05E7F50F374B_il2cpp_TypeInfo_var);
		Debug_Log_mC26E5AD0D8D156C7FFD173AA15827F69225E9DB8(_stringLiteral451AB6E913FABE3A79925211610F0CDBFD32D7BF, /*hidden argument*/NULL);
	}

IL_001d:
	{
		// }
		return;
	}
}
```

**Comment by JalonSolov** [comment]
Perhaps as an option, but I wouldn't want this by default.

Every line in the file must be parsed by the C compiler, even if it is comments, or even blank lines.

Every line parsed slows the compiler down, even if only by a "tiny" amount... and lots of "tiny" amounts can add up to a large, noticeable amount.

99% of the time, you don't need to look at the generated C file.  So having to do something extra to get this shouldn't be a huge issue.

**Comment by ntrel** [comment]
> Every line parsed slows the compiler down,

Parsing C comments is extremely efficient and would likely use a compiler intrinsic to find the end of the comment. The slowdown is negligible.

If this was a real problem, we would have to rewrite some of cgen because we already do things that are non optimal for the C compiler (including comments). The website IIRC actually said that the C generated is readable.

**Comment by JalonSolov** [comment]
Yes, I understand that the generated C is readable.  I've looked at it, and it is probably the cleanest generated C I've seen.

And yes, the C compiler parsing is usually highly optimized.

That still doesn't change the fact that it _would_ cause an extra amount of time... as well as taking up more of my disk space since V leaves all the generated C files lying around, now (instead of cleaning them up after a successful compile).

**Comment by medvednikov** [comment]
This can be done, definitely not by default, but behind a flag.

> The website IIRC actually said that the C generated is readable.

@ntrel what website were you referring to?
